README
Rubyish-Attribute
=================

This provide ruby-like syntax to build setter and getter in convenient way

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Rubyish::Attribute

If yor need any support, look for more information at:

    Project Repository:
    http://github.com/shelling/rubyish-attribute/tree/master

COPYRIGHT AND LICENCE

Copyright (C) 2008 by shelling, gugod

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Rubyish::Attribute.

0.05  Nov 17 2008
    - default import three functions, qw(:all) no longer required
    - example for meta programming

0.04  Nov 16 2008
    - refine doc
    - fix warn message
    - add example

0.03  Nov 14 2008
    - return undef and print err msg while invoking forbidden accessor

0.02  Nov 13 2008
    - correct the writing format of POD

0.01  Nov 12 2008
    - original version; created by h2xs 1.23 with options
        -X -n Rubyish::Attribute


t/lib/Animal.pm
package Animal;
use Rubyish::Attribute qw(:all);

attr_accessor( [qw(name)] );
attr_reader( [qw(color)] );
attr_writer( [qw(type)] );

sub new {
    my ($class, $args) = @_;
    my $self = bless {}, $class;
    for (keys %$args) {
        $self->{$_} = $args->{$_};
    }
    $self;
}

1;



example.pl
#!/usr/bin/env perl

use 5.010;

use warnings;
use strict;


use lib 'lib';

{
    package Animal;

    use Data::Dumper;
    use self;
    use Rubyish::Attribute;
    
    attr_accessor( [qw(name)] );
    attr_writer( [qw(age)] );
    attr_reader( [qw(color)] );

    sub new {
        my ($class, $args) = @_;
        my $self = bless {}, $class;
        for (keys %$args) { $self->{$_} = $args->{$_} }
        $self;
    }

    sub inspect { print Dumper(self) }

    sub add_attr_accessor {
        attr_accessor($_[1]);
        self;
    }

    1;
}

my $dogy = Animal->new({
    name  => "lucky",
    age   => 1,
    color => "white",
});

say "=== simple ===";

say $dogy->age; #=>  undef (with warn msg)
$dogy->age(2)->inspect;

say "\n****************************************";

say $dogy->color; #=> white
$dogy->color("black"); # undef (with warn msg)

say "\n****************************************";

say $dogy->name; #=> lucky
$dogy->name("jack")->inspect; 

say "\n=== meta programming ===";

$dogy->add_attr_accessor([qw(master)]);
$dogy->master("shelling")->inspect;

say "\n****************************************";



lib/Rubyish/Attribute.pm

=head1 NAME

Rubyish::Attribute - ruby-like accessor builder: attr_accessor, attr_writer and attr_reader.

=cut

package Rubyish::Attribute;

use Sub::Exporter;
Sub::Exporter::setup_exporter({ 
    exports => [qw(attr_accessor attr_writer attr_reader)],
    groups  => { default => [-all] },
});


=head1 VERSION

version 0.05

=cut

our $VERSION = "0.05";

=head1 SYNOPSIS

    #!/usr/bin/env perl
   
    use 5.010;

    use strict;
    use warnings;

    {
        package Animal;
        
        use Rubyish::Attribute; 
        # import attr_accessor, attr_writer and attr_reader

        attr_accessor( [qw(name color type)] ); 
        # pass an arrayref as the only one parameter

        # then create a constructer based on hashref
        sub new {
            $class = shift;
            bless {}, $class;
        }

        1;
    }
   
    $dogy = Animal->new()->name("rock")
                  ->color("black")->type("unknown");
    # new Animal with three attribute

    say $dogy->name;  #=> rock
    say $dogy->color; #=> black
    say $dogy->type;  #=> unknown

=head1 FUNCTIONS

=head2 attr_accessor(\@arrayref)

attr_accessor provides getters double as setters.
Because all setter return instance itself, now we can manipulate object in ruby way more than ruby.

    attr_accessor( [qw(name color type master)] )
    $dogy = Animal->new()->name("lucky")->color("white")
                  ->type("unknown")->master("shelling");

Each attribute could be read by getter as showing in synopsis.

=cut

sub attr_accessor {
    no strict;
    my $methods = shift;
    my $class = (caller)[0];

    my $make_accessor = sub {
        my $field = shift;
        return sub {
            my ($self, $arg) = @_;
            if ($arg) {
                $self->{$field} = $arg;
                $self;
            } else {
                $self->{$field};
            }
        }
    };

    for $field (@$methods) {
        *{$class . "::" . $field} = $make_accessor->($field);
    }
}

=head2 attr_reader(\@arrayref)

attr_reader create only getter for the class you call it

    attr_reader( [qw(name)] ) # pass an arrayref
    $dogy = Animal->new({name => "rock"}) # if we write initialize function in constructor
    $dogy->name()       #=> rock
    $dogy->name("jack") #=> undef (with warn msg)

=cut

sub attr_reader {
    no strict;
    my $methods = shift;
    my $class = (caller)[0];

    my $make_reader = sub {
        my $field = shift;
        return sub {
            my ($self, $arg) = @_;
            if ($arg) {
                warn "error - $field is only reader\n";
                return; # because no writer
            } else {
                $self->{$field};
            }
        }
    };
    
    for $field (@$methods) {
        *{$class . "::" . $field} = $make_reader->($field);
    }
}

=head2 attr_writer(\@arrayref)

attr_writer create only setter for the class you call it.

    attr_writer( [qw(name)] ) # pass an arrayref
    $dogy = Animal->new()->name("lucky") # initialize and set and get instance itself
    $dogy->name("jack") #=> instance itself 
    $dogy->name         #=> undef (with warn msg)

=cut

sub attr_writer {
    no strict;
    my $methods = shift;
    my $class = (caller)[0];

    my $make_writer = sub {
        my $field = shift;
        return sub {
            my ($self, $arg) = @_;
            if ($arg) {
                $self->{$field} = $arg;
                $self;
            } else {
                warn "error - $field is only writer\n";
                return; # because no reader 
            }
        }
    };

    for $field (@$methods) {
        *{$class . "::" . $field} = $make_writer->($field);
    }
}

=head1 DEPENDENCE

L<Sub::Exporter>

=head1 SEE ALSO

L<Sub::Exporter>, L<autobox::Core>, L<List::Rubyish>

L<http://ruby-doc.org/core-1.8.7/classes/Module.html#M000423>

L<http://chupei.pm.org/2008/11/rubyish-attribute.html> chinese introduction

=head1 AUTHOR

shelling <navyblueshellingford at gmail.com>

gugod    <gugod at gugod.org>

=head2 acknowledgement

Thanks to gugod providing testing script and leading me on the way of perl

=head1 REPOSITORY

host:       L<http://github.com/shelling/rubyish-attribute/tree/master>

checkout:   git clone git://github.com/shelling/rubyish-attribute.git

=head1 BUGS

please report bugs to <shelling at cpan.org> or <gugod at gugod.org>

=head1 COPYRIGHT & LICENCE 

Copyright © 2008 shelling, gugod, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1;

